File: typescript/src/content/media.ts
import { readFileSync, statSync } from 'node:fs';
import { posix } from 'node:path';
import { fileURLToPath } from 'node:url';

/** A GenAI content part carrying a payload, or recording one it could not carry. */
export type MediaPart =
  | { type: 'blob'; mime_type?: string; content: string }
  | { type: 'blob'; mime_type?: string; content_omitted: true }
  | { type: 'uri'; mime_type?: string; uri: string };

const DATA_URI = 'data:';
const FILE_URI = 'file:';
const REMOTE_SCHEMES = ['http', 'https'];

const MIME_BY_EXTENSION: Record<string, string> = {
  '.avif': 'image/avif',
  '.bmp': 'image/bmp',
  '.gif': 'image/gif',
  '.heic': 'image/heic',
  '.jpeg': 'image/jpeg',
  '.jpg': 'image/jpeg',
  '.pdf': 'application/pdf',
  '.png': 'image/png',
  '.svg': 'image/svg+xml',
  '.tif': 'image/tiff',
  '.tiff': 'image/tiff',
  '.webp': 'image/webp',
};

const PDF_MIME_TYPES = ['application/pdf', 'application/x-pdf'];

const MAX_PATH_LENGTH = 1024;
const MIN_BASE64_LENGTH = 32;
const BASE64_PATTERN = /^[A-Za-z0-9+/=\r\n]+$/;
const SCHEME_PATTERN = /^([a-zA-Z][a-zA-Z0-9+.-]*):/;

export type Bytes = Buffer | Uint8Array | ArrayBuffer;

export function isBytes(value: unknown): value is Bytes {
  return (
    value instanceof Uint8Array ||
    value instanceof ArrayBuffer ||
    Buffer.isBuffer(value)
  );
}

function scheme(uri: string): string {
  return SCHEME_PATTERN.exec(uri)?.[1]?.toLowerCase() ?? '';
}

function pathOf(uri: string): string {
  return uri.split('#')[0]!.split('?')[0]!;
}

function normalizeMime(value: unknown): string | undefined {
  if (typeof value !== 'string') return undefined;
  return value.split(';')[0]!.trim().toLowerCase() || undefined;
}

function mimeFromPath(uri: string): string | undefined {
  return MIME_BY_EXTENSION[posix.extname(pathOf(uri)).toLowerCase()];
}

function decodedLength(encoded: string): number {
  const body = encoded.trim();
  const padding = body.length - body.replace(/=+$/, '').length;
  return Math.max(Math.floor(body.length / 4) * 3 - padding, 0);
}

function isBase64(value: string): boolean {
  return (
    value.length >= MIN_BASE64_LENGTH &&
    value.length % 4 === 0 &&
    BASE64_PATTERN.test(value)
  );
}

/** Types the consumer can keep; others cost their whole payload for nothing. */
export function supportedMedia(mimeType: string | undefined): boolean {
  return (
    typeof mimeType === 'string' &&
    (mimeType.startsWith('image/') || PDF_MIME_TYPES.includes(mimeType))
  );
}

export interface MediaOptions {
  uri?: string | undefined;
  data?: Bytes | undefined;
  encoded?: string | undefined;
  mimeType?: string | undefined;
}

/** One non-text payload of a model call, read only when a part is built. */
export class Media {
  readonly mimeType: string | undefined;
  readonly uri: string | undefined;
  #data: Buffer | undefined;
  #encoded: string | undefined;
  #size: number | undefined;
  #unreadable = false;

  constructor({ uri, data, encoded, mimeType }: MediaOptions = {}) {
    this.uri = typeof uri === 'string' && uri ? uri : undefined;
    this.#data = isBytes(data) ? Buffer.from(data as Uint8Array) : undefined;
    this.#encoded =
      typeof encoded === 'string' && encoded ? encoded : undefined;
    this.mimeType =
      normalizeMime(mimeType) ??
      (this.uri === undefined ? undefined : mimeFromPath(this.uri));
  }

  static fromBytes(data: unknown, mimeType?: string): Media | undefined {
    if (!isBytes(data) || !new Uint8Array(data as Uint8Array).byteLength)
      return undefined;
    return new Media({ data: data as Bytes, mimeType });
  }

  static fromBase64(encoded: unknown, mimeType?: string): Media | undefined {
    if (typeof encoded !== 'string' || !encoded) return undefined;
    return new Media({ encoded, mimeType });
  }

  static fromUri(uri: unknown, mimeType?: string): Media | undefined {
    if (typeof uri !== 'string' || !uri) return undefined;
    return new Media({ uri, mimeType });
  }

  static fromDataUri(value: unknown): Media | undefined {
    if (typeof value !== 'string' || !value.startsWith(DATA_URI))
      return undefined;
    const separator = value.indexOf(',');
    if (separator < 0) return undefined;
    const header = value.slice(DATA_URI.length, separator);
    const payload = value.slice(separator + 1);
    if (!payload) return undefined;
    const parameters = header.split(';');
    if (!parameters.some((p) => p.trim().toLowerCase() === 'base64'))
      return undefined;
    return new Media({ encoded: payload, mimeType: parameters[0] });
  }

  /** Read an opaque value; prefer a constructor when the field is named. */
  static parse(value: unknown, mimeType?: string): Media | undefined {
    if (isBytes(value)) return Media.fromBytes(value, mimeType);
    if (typeof value !== 'string' || !value) return undefined;
    if (value.startsWith(DATA_URI)) return Media.fromDataUri(value);
    const prefix = scheme(value);
    if (REMOTE_SCHEMES.includes(prefix) || value.startsWith(FILE_URI))
      return Media.fromUri(value, mimeType);
    // A declared mime type is what separates raw base64 from a short path.
    if (mimeType !== undefined && isBase64(value))
      return Media.fromBase64(value, mimeType);
    if (value.length <= MAX_PATH_LENGTH) return Media.fromUri(value, mimeType);
    return undefined;
  }

  get isInline(): boolean {
    return this.#data !== undefined || this.#encoded !== undefined;
  }

  get isRemote(): boolean {
    // Any scheme we cannot open ourselves is somebody else's to fetch, so
    // gs:// and s3:// travel as references rather than being dropped.
    if (this.isInline || this.uri === undefined) return false;
    const prefix = scheme(this.uri);
    return prefix.length > 1 && prefix !== 'file';
  }

  get filename(): string | undefined {
    if (this.uri === undefined) return undefined;
    return posix.basename(pathOf(this.uri)) || undefined;
  }

  localPath(): string | undefined {
    if (this.isInline || this.uri === undefined || this.isRemote)
      return undefined;
    const prefix = scheme(this.uri);
    if (prefix.length > 1 && prefix !== 'file') return undefined;
    if (prefix !== 'file') return this.uri;
    try {
      return fileURLToPath(this.uri) || undefined;
    } catch {
      return undefined;
    }
  }

  /** Decoded size, without reading a file or decoding base64. */
  byteSize(): number | undefined {
    if (this.#size !== undefined) return this.#size;
    if (this.#data !== undefined) this.#size = this.#data.byteLength;
    else if (this.#encoded !== undefined)
      this.#size = decodedLength(this.#encoded);
    else {
      const path = this.localPath();
      if (path === undefined || this.#unreadable) return undefined;
      try {
        this.#size = statSync(path).size;
      } catch {
        this.#unreadable = true;
        return undefined;
      }
    }
    return this.#size;
  }

  base64(maxBytes?: number): string | undefined {
    if (maxBytes !== undefined) {
      const size = this.byteSize();
      if (size === undefined || size > maxBytes) return undefined;
    }
    if (this.#encoded !== undefined) return this.#encoded;
    if (this.#data === undefined) {
      const path = this.localPath();
      if (path === undefined || this.#unreadable) return undefined;
      try {
        this.#data = readFileSync(path);
      } catch {
        this.#unreadable = true;
        return undefined;
      }
      this.#size = this.#data.byteLength;
    }
    this.#encoded = this.#data.toString('base64');
    return this.#encoded;
  }

  toPart(maxBytes?: number): MediaPart {
    const mime =
      this.mimeType === undefined ? {} : { mime_type: this.mimeType };
    // A reference costs nothing to carry whatever its type.
    if (this.isRemote) return { type: 'uri', ...mime, uri: this.uri! };
    const encoded = supportedMedia(this.mimeType)
      ? this.base64(maxBytes)
      : undefined;
    return encoded === undefined
      ? { type: 'blob', ...mime, content_omitted: true }
      : { type: 'blob', ...mime, content: encoded };
  }

  toString(): string {
    return `Media(mimeType=${this.mimeType ?? '?'}, source=${
      this.uri ?? 'inline'
    }, bytes=${this.#size ?? '?'})`;
  }
}

File: python/src/confident_trace/_core/media.py
"""Media normalized out of provider payloads into GenAI content parts."""

from __future__ import annotations

from base64 import b64encode
from os import stat
from os.path import basename, splitext
from urllib.parse import unquote, urlparse

from .safety import safe

_DATA_URI = "data:"
_FILE_URI = "file:"
_REMOTE_SCHEMES = ("http://", "https://")

_MIME_BY_EXTENSION = {
    ".avif": "image/avif",
    ".bmp": "image/bmp",
    ".gif": "image/gif",
    ".heic": "image/heic",
    ".jpeg": "image/jpeg",
    ".jpg": "image/jpeg",
    ".pdf": "application/pdf",
    ".png": "image/png",
    ".svg": "image/svg+xml",
    ".tif": "image/tiff",
    ".tiff": "image/tiff",
    ".webp": "image/webp",
}

_PDF_MIME_TYPES = ("application/pdf", "application/x-pdf")

_MAX_PATH_LENGTH = 1024
_MIN_BASE64_LENGTH = 32
_BASE64_CHARS = frozenset(
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=\r\n"
)


def _read(path):
    with open(path, "rb") as handle:
        return handle.read()


def _file_size(path):
    return stat(path).st_size


def _mime(value):
    if type(value) is not str:
        return None
    return value.split(";", 1)[0].strip().lower() or None


def _mime_from_path(path):
    return _MIME_BY_EXTENSION.get(splitext(urlparse(path).path or path)[1].lower())


def _decoded_length(encoded):
    body = encoded.strip()
    padding = len(body) - len(body.rstrip("="))
    return max(len(body) // 4 * 3 - padding, 0)


def supported_media(mime_type):
    return type(mime_type) is str and (
        mime_type.startswith("image/") or mime_type in _PDF_MIME_TYPES
    )


def _is_base64(value):
    return (
        len(value) >= _MIN_BASE64_LENGTH
        and len(value) % 4 == 0
        and set(value) <= _BASE64_CHARS
    )


class Media:
    """One non-text payload of a model call, read only when a part is built."""

    __slots__ = ("mime_type", "uri", "_data", "_encoded", "_size", "_unreadable")

    def __init__(self, *, uri=None, data=None, encoded=None, mime_type=None):
        self.uri = uri if type(uri) is str and uri else None
        if type(data) is bytes:
            self._data = data
        elif isinstance(data, (bytearray, memoryview)):
            self._data = bytes(data)
        else:
            self._data = None
        self._encoded = encoded if type(encoded) is str and encoded else None
        self.mime_type = _mime(mime_type) or (
            _mime_from_path(self.uri) if self.uri else None
        )
        self._size = None
        self._unreadable = False

    @classmethod
    def from_bytes(cls, data, mime_type=None):
        if not isinstance(data, (bytes, bytearray, memoryview)) or not data:
            return None
        return cls(data=data, mime_type=mime_type)

    @classmethod
    def from_base64(cls, encoded, mime_type=None):
        if type(encoded) is not str or not encoded:
            return None
        return cls(encoded=encoded, mime_type=mime_type)

    @classmethod
    def from_uri(cls, uri, mime_type=None):
        if type(uri) is not str or not uri:
            return None
        return cls(uri=uri, mime_type=mime_type)

    @classmethod
    def from_data_uri(cls, value):
        if type(value) is not str or not value.startswith(_DATA_URI):
            return None
        header, separator, payload = value[len(_DATA_URI) :].partition(",")
        if not separator or not payload:
            return None
        parameters = header.split(";")
        if "base64" not in (parameter.strip().lower() for parameter in parameters):
            return None
        return cls(encoded=payload, mime_type=parameters[0])

    @classmethod
    def parse(cls, value, mime_type=None):
        """Read an opaque value; prefer a constructor when the field is named."""
        if isinstance(value, (bytes, bytearray, memoryview)):
            return cls.from_bytes(value, mime_type)
        if type(value) is not str or not value:
            return None
        if value.startswith(_DATA_URI):
            return cls.from_data_uri(value)
        if value.startswith(_REMOTE_SCHEMES) or value.startswith(_FILE_URI):
            return cls.from_uri(value, mime_type)
        # A declared mime type is what separates raw base64 from a short path.
        if mime_type is not None and _is_base64(value):
            return cls.from_base64(value, mime_type)
        if len(value) <= _MAX_PATH_LENGTH:
            return cls.from_uri(value, mime_type)
        return None

    @property
    def is_inline(self):
        return self._data is not None or self._encoded is not None

    @property
    def is_remote(self):
        if self.is_inline or self.uri is None:
            return False
        scheme = urlparse(self.uri).scheme
        return len(scheme) > 1 and scheme != "file"

    @property
    def filename(self):
        if self.uri is None:
            return None
        return basename(urlparse(self.uri).path) or None

    def local_path(self):
        if self.is_inline or self.uri is None or self.is_remote:
            return None
        parsed = urlparse(self.uri)
        # A single-character scheme is a Windows drive letter, not a scheme.
        if len(parsed.scheme) > 1 and parsed.scheme != "file":
            return None
        if parsed.scheme != "file":
            return self.uri
        raw = f"//{parsed.netloc}{parsed.path}" if parsed.netloc else parsed.path
        return unquote(raw) or None

    def byte_size(self):
        """Decoded size, without reading a file or decoding base64."""
        if self._size is not None:
            return self._size
        if self._data is not None:
            self._size = len(self._data)
        elif self._encoded is not None:
            self._size = _decoded_length(self._encoded)
        else:
            path = self.local_path()
            if path is None or self._unreadable:
                return None
            size = safe(_file_size, path)
            if size is None:
                self._unreadable = True
                return None
            self._size = size
        return self._size

    def base64(self, max_bytes=None):
        if max_bytes is not None:
            size = self.byte_size()
            if size is None or size > max_bytes:
                return None
        if self._encoded is not None:
            return self._encoded
        if self._data is None:
            path = self.local_path()
            if path is None or self._unreadable:
                return None
            data = safe(_read, path)
            if data is None:
                self._unreadable = True
                return None
            self._data = data
            self._size = len(data)
        self._encoded = b64encode(self._data).decode("ascii")
        return self._encoded

    def to_part(self, max_bytes=None):
        part = {"type": "uri" if self.is_remote else "blob"}
        if self.mime_type is not None:
            part["mime_type"] = self.mime_type
        if self.is_remote:
            # A reference costs nothing to carry whatever its type.
            part["uri"] = self.uri
            return part
        encoded = self.base64(max_bytes) if supported_media(self.mime_type) else None
        if encoded is None:
            part["content_omitted"] = True
        else:
            part["content"] = encoded
        return part

    def __repr__(self):
        # Reprs reach logs; the payload must never ride along.
        source = self.uri if self.uri is not None else "inline"
        size = "?" if self._size is None else self._size
        return f"Media(mime_type={self.mime_type!r}, source={source!r}, bytes={size})"
